### ISSUE TITLE
Enforce consistent `return` statements

### ISSUE THREAD

**Issue by sobolevn** [wemake-services/wemake-python-styleguide#378]
# Rule request

## Question

There are three ways to `return None` in python:

1. `return`
2. `return None`
3. function without a `return`

We need to ask developers what do they think about this inconsistency. 
My opinion is that we should only keep `2` and possibly `3`. 

Why keeping `2`? It is the most explicit notation out of the two options.

Why possibly `3`? In my ideal world all functions should return something. Ideally in one place.
Yearly returns are also fine, but in this case all branches of this function must return something explicitly. 

## Thesis

After several hours of thinking about this problem and asking opinions of other developers, this is what I came to:

1. If there are no returns in a functions - that's fine
2. If not all code paths return - that's fine
3. If all `return` in a functions uses short form, that's fine
4. If any `return` in a function uses explicit form `return something`, it means that all `return`s must be explicit: `return something`

Related: https://eslint.org/docs/rules/consistent-return

## Reasoning

This is done for consistency reasons.

**Comment by sobolevn** [comment]
## Examples

### Correct

```python
def test(arg: int) -> int:
    if arg > 0:
        return arg
    return -arg
```

```python
def test(arg: int) -> None:
    if arg > 0:
        return
    if arg < 10:
        return
    some_call(arg)
```


```python
def test(arg: int) -> None:
    if arg > 0:
        return arg
    assert arg < 10  # no-return
```

### Wrong

```python
def test(arg: int) -> None:
    if arg > 0:
        return
    return None
```

```python
def test(arg: int) -> None:
    if arg > 0:
        return None
    return
```

**Comment by kxepal** [comment]
```
def test(arg: int) -> None:
    if arg > 0:
        return arg
    assert arg < 10  # no-return
```
Is that correct?

**Comment by kxepal** [comment]
:beer: 
However, examples two and three quite simiar. May be improve second one with `return None` in both of cases?

**Comment by sobolevn** [comment]
Ok, after some initial attempts to handle this I have found the right explanation for this feature. It consists of several independent rules.

## Consistent return values

We only check that if any return contains a value - all returns must contain a value. We do not check types or that all path returns. Only the value.

## Last line extra return

It is possible to finish your function with a `return` statement. Like so:

```python
def some():
      ...
      return
```

Or

```python
def some():
      ...
      return some_value_or_None
```

The thing is: just `return` in the last line is not required at all. It is more readable not to write it.
But having `return some_value_or_None` is always possible.

**Comment by kxepal** [comment]
@sobolevn 
Actually, I'm not sure how to start with all these asts. Could you recommend any good example around?

**Comment by sobolevn** [comment]
A good place to start is https://wemake-python-styleguide.readthedocs.io/en/latest/pages/api.html
Then head to https://wemake-python-styleguide.readthedocs.io/en/latest/pages/contributing.html and https://greentreesnakes.readthedocs.io/en/latest/

Then read about what `visitors` are: https://wemake-python-styleguide.readthedocs.io/en/latest/pages/visitors/base.html
Then about `violations`: https://wemake-python-styleguide.readthedocs.io/en/latest/pages/violations/base.html

You will need to create a new `visitor` in this file: https://github.com/wemake-services/wemake-python-styleguide/blob/master/wemake_python_styleguide/visitors/ast/keywords.py
And a new `violation` here: https://github.com/wemake-services/wemake-python-styleguide/blob/master/wemake_python_styleguide/violations/consistency.py

**Comment by kxepal** [comment]
Thanks! I'm going to grab some tea for this.

**Comment by sobolevn** [comment]
@kxepal how's your progress? Should I take this over or not?

**Comment by kxepal** [comment]
Better don't count on me here. But I'm still watching the changes to catch the wave.

**Comment by sobolevn** [comment]
`pylint` actually has this check: https://github.com/PyCQA/pylint/blob/5a86fddcb815ebd08fbc3ffcbaf7b11c2ba2e107/pylint/checkers/refactoring.py#L128

So, it may be easier to implement.

**Comment by kxepal** [comment]
Wow, never saw that one.

**Comment by sobolevn** [comment]
@kxepal I guess consistent values rule is also the case with `yield`.

```python
def function():
      if some:
           yield
     yield None
     yield 1
```

I see no difference with `return` in how values should be defined.
